colcon_cmake/event_handler/compile_commands.py

```diff
@@ -0,0 +1,121 @@
+# Copyright 2020 Dirk Thomas
+# Licensed under the Apache License, Version 2.0
+
+import os
+from pathlib import Path
+
+from colcon_core.event.job import JobQueued
+from colcon_core.event.job import JobUnselected
+from colcon_core.event_handler import EventHandlerExtensionPoint
+from colcon_core.event_reactor import EventReactorShutdown
+from colcon_core.logging import colcon_logger
+from colcon_core.plugin_system import satisfies_version
+
+logger = colcon_logger.getChild(__name__)
+
+
+class CompileCommandsEventHandler(EventHandlerExtensionPoint):
+    """
+    Generate a `compile_commands.json` file for the whole workspace.
+
+    The file is created in the build directory and aggregates the data from all
+    packages.
+    """
+
+    FILENAME = 'compile_commands.json'
+
+    # the priority should be lower than e.g. the status and summary extensions
+    # in order to show those as soon as possible
+    PRIORITY = 40
+
+    def __init__(self):  # noqa: D107
+        super().__init__()
+        satisfies_version(
+            EventHandlerExtensionPoint.EXTENSION_POINT_VERSION, '^1.0')
+        self._package_names = set()
+
+    def __call__(self, event):  # noqa: D102
+        data = event[0]
+
+        if isinstance(data, JobQueued) or isinstance(data, JobUnselected):
+            # delay loading json for all packages
+            self._package_names.add(data.identifier)
+
+        elif isinstance(data, EventReactorShutdown):
+            # if no package has a json file there is no need for a
+            # workspace-level json file
+            package_level_json_paths = self._get_package_level_json_paths()
+            workspace_level_json_path = self._get_path()
+            if not package_level_json_paths:
+                if workspace_level_json_path.exists():
+                    workspace_level_json_path.unlink()
+                return
+
+            # if the workspace-level json file is newer than all package-level
+            # json files it doesn't need to be regenerated
+            try:
+                workspace_level_mtime = os.path.getmtime(
+                    str(workspace_level_json_path))
+            except Exception:
+                pass
+            else:
+                for json_path in sorted(package_level_json_paths):
+                    try:
+                        mtime = os.path.getmtime(str(json_path))
+                    except Exception:
+                        continue
+                    if mtime > workspace_level_mtime:
+                        break
+                else:
+                    return
+
+            # collect all package-level json data
+            # for performance reasons avoid to load/dump json data
+            all_compile_commands = []
```

Comment:
This could get pretty big, couldn't it? Would it be worth it to relay the compile commands directly to the destination file as they are read instead of buffering them all together?

Reply:
The overall size of the combined json file in a ROS 2 workspace (`ros2.repos / ~300 pkgs) is around 7.5 MB.

I can rework the logic to write each `compile_commands` right after it was read without keeping all in memory if you think that makes sense.

Reply:
As long as you've given it some thought, I'm fine with leaving it as-is. Thanks for getting some tangible data on it.

Reply:
Actually the original logic comes from when the code was collecting all compile commands in a list and then used `yaml.dump` to write the workspace-level file. Since it doesn't do that anymore there is not need to keep all compile command in memory. I refactored it in 9debce6b5ccdb726b541e2b513e8c6e3fae87ba4.